test/upgrade/dynamic/service.go

```diff
@@ -0,0 +1,283 @@
+	//Choosing resource version dynamically
+	gvr := GetGroupVersionResource("")
+
+	service, err := clients.Dynamic.Resource(gvr).Namespace(test.ServingNamespace).Get(serviceName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get %s: %#v", serviceName, err)
+	}
+	resource, err := populateResources(*service)
+	if err != nil {
+		t.Fatalf("Failed to populate resource: %#v", err)
+	}
+	return resource
+}
+
+//PatchServiceImage patches the existing service passed in with a new newImage. Returns the latest test.ResourceNames
+func PatchServiceImage(t *testing.T, clients *test.Clients, serviceName string, newImage string) test.ResourceNames {
+	//Choosing resource version dynamically
+	gvr := GetGroupVersionResource("")
+	var patchedBytes []byte
+	service, err := clients.Dynamic.Resource(gvr).Namespace(test.ServingNamespace).Get(serviceName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get %s: %#v", serviceName, err)
+	}
+
+	if v1beta1.SchemeGroupVersion.String() == service.GetAPIVersion() {
+		serviceObject := v1beta1.Service{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(service.UnstructuredContent(), &serviceObject); err != nil {
+			t.Fatalf("Failed to convert to object: %#v", err)
+		}
+		newServiceObject := serviceObject.DeepCopy()
+		v1b1testing.WithServiceImage(newImage)(newServiceObject)
+		patchedBytes, err = createPatch(serviceObject, newServiceObject)
+		if err != nil {
+			t.Fatalf("Failed to get bytes : %#v", err)
+		}
+	} else if v1alpha1.SchemeGroupVersion.String() == service.GetAPIVersion() {
+		serviceObject := v1alpha1.Service{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(service.UnstructuredContent(), &serviceObject); err != nil {
+			t.Fatalf("Failed to convert to object: %#v", err)
+		}
+		newServiceObject := serviceObject.DeepCopy()
+		v1a1test.SetServiceImage(newServiceObject, newImage)
+		patchedBytes, err = createPatch(serviceObject, newServiceObject)
+		if err != nil {
+			t.Fatalf("Failed to get bytes : %#v", err)
+		}
+	}
+
+	gvr = GetGroupVersionResource(service.GetAPIVersion())
+	patchedService, err := clients.Dynamic.Resource(gvr).Namespace(test.ServingNamespace).Patch(serviceName, types.JSONPatchType, patchedBytes, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to patch %s: %#v", serviceName, err)
+	}
+	resource, err := populateResources(*patchedService)
+	if err != nil {
+		t.Fatalf("Failed to populate resource: %#v", err)
+	}
+	return resource
+}
+
+// WaitForServiceState polls the status of the Service called serviceName
+// from client every `interval` until `inState` returns `true` indicating it
+// is done, returns an error or timeout. desc will be used to name the metric
+// that is emitted to track how long it took for name to get into the state checked by inState.
+func WaitForServiceState(t *testing.T, clients *test.Clients, serviceName string, desc string, inState func(svc *unstructured.Unstructured) (bool, error)) {
```

Comment:
Is the `bool` necessary as part of the return? Can we just check if err is nil like the current `WaitForServiceState`?